### ISSUE TITLE
.c3d marker offsets

### ISSUE THREAD

**Issue by NicholasMilef** [sha2nkt/moyo_toolkit#7]
Thanks for releasing this dataset!

I have a question regarding the .c3d files. Some of the markers have large offsets from the actual subject, which I've seen in some other .c3d datasets such as HumanEva. I have two questions regarding this:

1. Do you know what causes these offsets?
2. Are the markers filtered before being passed through MoSh++? 

Single frame:
![moshpp](https://github.com/sha2nkt/moyo_toolkit/assets/48165112/499edb9a-9302-45bf-b19f-838baac174e4)

Animation:
![moshpp_compressed (1)](https://github.com/sha2nkt/moyo_toolkit/assets/48165112/fd48d6c9-c5d4-4b5e-a690-434069b1ffb9)

**Comment by gfgb** [comment]
Hi Nicholas,
I tried to load the marker data myself and I wasn't able to reproduce this issue, they are well aligned for me.
Do you encounter this problem with all of the sequences or only specific ones?
Can you share the code you are using to place the markers into the scene?

Giorgio

**Comment by NicholasMilef** [comment]
Hi Giorgio,

Thanks for your help! I encountered this problem with multiple files (I'm not sure if every file has this), but I might be doing something wrong with the marker loading. I rendered the above images in Blender, but I wrote a script that I attached below with Pyrender that produces the same results. For the first sequence (Boat Pose), there were 202 markers in the C3D data, but there are obviously less as shown in the video.

```
import ezc3d
import imageio
import joblib
import numpy as np
import pyrender
import smplx
import trimesh
import torch
import os

anim_name = "221004_yogi_nexus_body_hands_03596_Boat_Pose_or_Paripurna_Navasana_-a"
#anim_name = "221004_yogi_nexus_body_hands_03596_Low_Lunge_pose_or_Anjaneyasana_-d"
#anim_name = "221004_yogi_nexus_body_hands_03596_Standing_Split_pose_or_Urdhva_Prasarita_Eka_Padasana_-a"
smplx_filename = "data/MOYO/20221004_with_com/mosh/train/" + anim_name + "_stageii.pkl"
c3d_filename = "data/MOYO/20221004_with_com/pressure/train/pressure_mat_c3d/" + anim_name + ".c3d"
body_model_path = "<path_to_body_models>"  # path to SMPL-X body models

# setup scene
zoom = 0.75
camera_matrix = [
    [0.750, -0.252, 0.611, 3.4 * zoom],
    [0.661, 0.291, -0.691, -2.0 * zoom],
    [-0.004, 0.923, 0.385, 2.3 * zoom],
    [0, 0, 0, 1],
]

scene = pyrender.Scene(ambient_light=[0.2, 0.2, 0.2])

camera = pyrender.PerspectiveCamera(yfov=np.pi / 3.0)
scene.add(camera, pose=camera_matrix)

renderer = pyrender.OffscreenRenderer(640, 480, 1.0)

# create floor mesh
floor_mesh = trimesh.Trimesh(
    vertices=[[-25, -25, 0], [-25, 25, 0], [25, -25, 0], [25, 25, 0]],
    faces=[[0, 2, 1], [1, 2, 3]],
)
f_node = pyrender.Mesh.from_trimesh(floor_mesh)
f_node = scene.add(f_node)

# setup lighting
directional_light = pyrender.DirectionalLight(color=[1.0, 1.0, 1.0], intensity=5.0)
scene.add(directional_light)

# load markers
c3d_freq = ezc3d.c3d(c3d_filename)["parameters"]["POINT"]["RATE"]["value"].item()
c3d_points = np.nan_to_num(ezc3d.c3d(c3d_filename)["data"]["points"], 0)  # [4, M, F]
c3d_points /= 1000  # convert from millimeters to meters
c3d_points = c3d_points.transpose((2, 1, 0))[:, :, :3]  # [4, M, F] -> [F, M, 3]
sm = trimesh.creation.uv_sphere(radius=0.02)
sm.visual.vertex_colors = [1.0, 0.0, 0.0]
tfs = np.repeat(np.expand_dims(np.eye(4), axis=0), axis=0, repeats=c3d_points.shape[1])
c3d_mesh = pyrender.Mesh.from_trimesh(sm, poses=tfs)
c3d_node = scene.add(c3d_mesh)

# load smplx meshes
smplx_data = joblib.load(smplx_filename)
num_frames = smplx_data["fullpose"].shape[0]

smplx_model = smplx.create(
    model_path=body_model_path,
    model_type="smplx",
    num_betas=10,
    gender="neutral",
    batch_size=num_frames,
)

poses = torch.from_numpy(smplx_data["fullpose"]).float()
betas = torch.from_numpy(smplx_data["betas"]).unsqueeze(0).float()
betas = torch.repeat_interleave(betas, dim=0, repeats=num_frames)
transl = torch.from_numpy(smplx_data["trans"]).float()

vertices = smplx_model(
    global_orient=poses[:, :3],
    body_pose=poses[:, 3:66],
    betas=betas[:, :10],
    transl=transl,
)["vertices"].detach().cpu().numpy()

smplx_meshes = []
for frame in range(num_frames):
    tm = trimesh.Trimesh(
        vertices=vertices[frame],
        faces=smplx_model.faces,
    )
    tm.visual.vertex_colors = [0.3, 0.5, 1.0]
    smplx_meshes.append(pyrender.Mesh.from_trimesh(tm))
smplx_node = scene.add(smplx_meshes[0])


# render loop
def render_frame(frame):
    # update markers
    tfs[:, :3, 3] = c3d_points[frame]
    c3d_node.mesh = pyrender.Mesh.from_trimesh(sm, poses=tfs)

    # update SMPLX meshes
    smplx_node.mesh = smplx_meshes[frame]


# write out .gif
video_freq = 30.0
output_filename = os.path.basename(c3d_filename).replace(".c3d", ".gif")
with imageio.get_writer(output_filename, mode="I", duration=(1000.0 * (1.0 / video_freq))) as writer:
    for frame in range(0, num_frames, int(c3d_freq / video_freq)):
        render_frame(frame)
        color, _ = renderer.render(scene, flags=pyrender.constants.RenderFlags.SHADOWS_ALL)
        writer.append_data(color)

```

![221004_yogi_nexus_body_hands_03596_Standing_Split_pose_or_Urdhva_Prasarita_Eka_Padasana_-a_compressed](https://github.com/sha2nkt/moyo_toolkit/assets/48165112/570fe7b3-c1c4-49b0-8660-0e5bc4cf7689)
![221004_yogi_nexus_body_hands_03596_Low_Lunge_pose_or_Anjaneyasana_-d_compressed](https://github.com/sha2nkt/moyo_toolkit/assets/48165112/fb3e5e23-fbcc-462b-a053-f9816c1715db)
![221004_yogi_nexus_body_hands_03596_Boat_Pose_or_Paripurna_Navasana_-a_compressed](https://github.com/sha2nkt/moyo_toolkit/assets/48165112/b32a664b-f0de-4352-8a75-7648ae5fe5b1)

**Comment by gfgb** [comment]
Hi Nicholas,
The c3d files of the "com" capture (those whose filename starts with "221004") do in fact contain more markers than they should. Before I only checked the files from the "vicon" capture and did not notice the issue.

I am not sure why these files contain this extra data, but luckily the solution is straightforward: the marker layout is made of 39 markers, so you can just select the first 39 markers in the c3d file. In your code, I think adding the line `c3d_points = c3d_points[:, :39, :] should to the trick.

In case for any reason they are not ordered correctly, these are the labels that you should select:
```
marker_names = ['LFHD', 'RFHD', 'LBHD', 'RBHD', 'C7', 'T10', 'CLAV', 'STRN', 'RBAK', 'LSHO', 'LUPA', 'LELB', 'LFRM',
                 'LWRA', 'LWRB', 'LFIN', 'RSHO', 'RUPA', 'RELB', 'RFRM', 'RWRA', 'RWRB', 'RFIN', 'LASI', 'RASI', 'LPSI',
                 'RPSI', 'LTHI', 'LKNE', 'LTIB', 'LANK', 'LHEE', 'LTOE', 'RTHI', 'RKNE', 'RTIB', 'RANK', 'RHEE', 'RTOE']
```
Let me know if this solves the problem.

FYI, only the markers listed above were used during the moshing process.

**Comment by NicholasMilef** [comment]
Thanks @gfgb! It looks good now!
![221004_yogi_nexus_body_hands_03596_Boat_Pose_or_Paripurna_Navasana_-a_compressed (1)](https://github.com/sha2nkt/moyo_toolkit/assets/48165112/7dabafcf-19f6-4649-b0d8-99ff2acb5cf2)

 Do you know what causes the other markers to appear?

**Comment by gfgb** [comment]
I suspect it could be due to some export setting in Vicon Shogun, the software that was used to clean the marker data. I will check with our data team.
